### DISCUSSION TITLE
Why aiohttp makes graphql-server Hasura slow?

### DISCUSSION THREAD

**Discussion by V-ampire** [aio-libs/aiohttp#7246] [Q&A]
Hi there! I use Hasura as graphql server as docker container. Also I use aiohttp to request hasura from another docker container and get a very slow RPS and actually hasura query execution time is big also. I discovered that the most of time takes parsing and compiling (not query execution) of SQL. In the same time I tested Hasura using for requests Locust which makes requests via greenlets and it provides much more higher RPS (about 2x). 

Could you please help me to understand why Hasura is so slow with aiohttp?

**Comment by Dreamsorcerer** [comment]
That's rather lacking in information. Can you demonstrate what your code looks like, and exactly which parts of the code are slowest? You say that most time is spent in parsing/compiling of SQL, which has nothing to do with aiohttp...

**Reply by V-ampire** [reply]
Sure, I supposed maybe aiohttp and graphql-server could share any resources..

This is my test code:

```
import asyncio
import time
import uvloop

import aiohttp
from aiohttp import ClientTimeout

uvloop.install()


async def execute_query(session, N):
    query = """
    query myQuery {.....}
    """
    payload = dict(query=query)
    async with session.post('http://hasura:8080/v1/graphql/', json=payload, ssl=False) as resp:
        data = await resp.json()
        if not data.get('data'):
            print(f"No data {data}")
        return data

async def send_requests(N: int):
    async with aiohttp.ClientSession(
        connector=aiohttp.TCPConnector(limit=9000),
        headers={'x-hasura-admin-secret': 'admin-secret'},
    ) as session:
        return await asyncio.gather(*[
            execute_query(session, N) for i in range(N)
        ])


def test_rpc(N):
    start_timestamp = time.time()

    asyncio.run(send_requests(N))

    task_time = round(time.time() - start_timestamp, 2)
    rps = round(N / task_time, 1)
    print(
        f"Requests: {N}; Total time: {task_time} s; RPS: {rps}."
    )


if __name__ == '__main__':
    test_rpc(1000)
    test_rpc(2000)
    test_rpc(5000)
    test_rpc(10000)
```

Results:
![image](https://user-images.githubusercontent.com/34309587/229141785-28629bdd-95de-49e5-af34-0753d087278c.png)

Actually locust results:
![image](https://user-images.githubusercontent.com/34309587/229141905-c33a98f1-b45f-4ae8-a71f-87ef953cb64d.png)

**Reply by Dreamsorcerer** [reply]
Code looks correct to me. Presumably graphql-server is running in another process, so there shouldn't be any resource sharing that would affect the client.

If you are seeing more CPU usage in SQL compilation, then that surely has nothing to do with aiohttp, unless it's sending the SQL in a different format or something which causes it to spend longer parsing...

Based on your example, the only thing I can think of trying is to remove uvloop, just in case it's causing something to slow down (sure it's meant to faster in most cases, but maybe you have an edge case...). If that has no effect, then it looks like we'd need more debugging information to figure out what's happening.

**Reply by V-ampire** [reply]
Thanks for reply, I tried to use asyncio event loop instead uvloop, no effects...
